src/main/resources/migration/changelog-v5.6.0.xml

```diff
@@ -107,4 +107,202 @@
                                  onDelete="NO ACTION" onUpdate="NO ACTION" referencedColumnNames="ID"
                                  referencedTableName="TAG" validate="true"/>
     </changeSet>
+
+    <changeSet id="v5.6.0-7" author="n1ckl0sk0rtge">
+        <sql splitStatements="true">
+            ALTER TABLE "COMPONENT" DROP CONSTRAINT IF EXISTS "COMPONENT_CLASSIFIER_check";
+            ALTER TABLE "COMPONENT" ADD CONSTRAINT "COMPONENT_CLASSIFIER_check"
+                CHECK ("CLASSIFIER" IS NULL OR "CLASSIFIER"::TEXT = ANY(ARRAY['APPLICATION', 'CONTAINER', 'DEVICE', 'FILE', 'FIRMWARE', 'FRAMEWORK', 'LIBRARY', 'OPERATING_SYSTEM', 'CRYPTOGRAPHIC_ASSET']));
+
+            ALTER TABLE "PROJECT" DROP CONSTRAINT IF EXISTS "PROJECT_CLASSIFIER_check";
+            ALTER TABLE "PROJECT" ADD CONSTRAINT "PROJECT_CLASSIFIER_check"
+                CHECK ("CLASSIFIER" IS NULL OR "CLASSIFIER"::TEXT = ANY(ARRAY['APPLICATION', 'CONTAINER', 'DEVICE', 'FILE', 'FIRMWARE', 'FRAMEWORK', 'LIBRARY', 'OPERATING_SYSTEM', 'CRYPTOGRAPHIC_ASSET']));
+        </sql>
+    </changeSet>
+
+    <changeSet id="v5.6.0-8" author="san-zrl">
+        <addColumn tableName="COMPONENT">
+            <column name="CRYPTO_PROPERTIES_ID" type="BIGINT"/>
+        </addColumn>
+
+        <createTable tableName="CRYPTO_PROPERTIES">
+            <column autoIncrement="true" name="ID" type="BIGINT">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="CRYPTO_PROPERTIES_PK"/>
+            </column>
```

Comment:
Yes, true. This would be a better idea but I didn't manage to make this work.

I tried to specify a COMPONENT_ID in CRYPTO_PROPERTIES that refers upwards to the enclosing component along with an FK constraint that would cause the right order of cascading deletes. This led to a problem when trying to persist the modified crypto asset properties object. I reproduced the error msg with stacktrace and appended it below for reference. It turns out that CryptoAssetProperties@5af49362 **is not the instance from ModelConverter** (in which the component id was properly set) but another instance created somewhere deep down in the transaction code.

```
2024-10-21 15:22:19,384 ERROR [Persist] Insert of object "org.dependencytrack.model.CryptoAssetProperties@5af49362" using statement "INSERT INTO "CRYPTO_PROPERTIES" ("ALGORITHM_PROPERTIES_ID","ASSET_TYPE","CERTIFICATE_PROPERTIES_ID","COMPONENT_ID","OID","PROTOCOL_PROPERTIES_ID","RELATED_MATERIAL_PROPERTIES_ID") VALUES (?,?,?,?,?,?,?)" failed : ERROR: null value in column "COMPONENT_ID" of relation "CRYPTO_PROPERTIES" violates not-null constraint
  Detail: Failing row contains (3, null, ALGORITHM, 3, null, null, null, 2.16.840.1.101.3.4.1.6). [bomSerialNumber=e8c355aa-2142-4084-a8c7-6d42c8610ba2, bomFormat=CycloneDX, bomUploadToken=6cb46db7-e009-4146-8615-122179457b63, projectName=x, bomSpecVersion=1.6, projectUuid=416f54cc-8621-492f-838f-b87ef3de9dad, projectVersion=null, bomVersion=1]
javax.jdo.JDODataStoreException: Insert of object "org.dependencytrack.model.CryptoAssetProperties@5af49362" using statement "INSERT INTO "CRYPTO_PROPERTIES" ("ALGORITHM_PROPERTIES_ID","ASSET_TYPE","CERTIFICATE_PROPERTIES_ID","COMPONENT_ID","OID","PROTOCOL_PROPERTIES_ID","RELATED_MATERIAL_PROPERTIES_ID") VALUES (?,?,?,?,?,?,?)" failed : ERROR: null value in column "COMPONENT_ID" of relation "CRYPTO_PROPERTIES" violates not-null constraint
  Detail: Failing row contains (3, null, ALGORITHM, 3, null, null, null, 2.16.840.1.101.3.4.1.6).
	at org.datanucleus.api.jdo.JDOAdapter.getJDOExceptionForNucleusException(JDOAdapter.java:608)
	at org.datanucleus.api.jdo.JDOPersistenceManager.flush(JDOPersistenceManager.java:2057)
	at org.dependencytrack.tasks.BomUploadProcessingTask.processComponents(BomUploadProcessingTask.java:529)
...
```

Reply:
Just to be sure, is this how you modeled the relationship in this case? https://www.datanucleus.org/products/accessplatform_6_0/jdo/mapping.html#one_many_fk_bi

Reply:
I wanted the 'inner' dependent object to be fetched when the 'outer' object is queried and I wanted the cascading delete to work in both ways. The code does that, so I think we can leave it as is.

The fetching works since the `inner` objects (e.g., `CryptoAssetProperties`) are modelled as persistent members of the `outer` (e.g., ` Component`) object. Inner objects cannot exist without their 'outer' container. Thus, cascading delete must work in both ways:
1. Deleting an 'outer' object must delete all 'inner' objects. This is guaranteed by the `dependent = "true"` flag in the `Persistent` annotation. Here is an example:  https://github.com/DependencyTrack/hyades-apiserver/blob/279450cf3af0bec1ff23e712a2168b5d3508250b/src/main/java/org/dependencytrack/model/Component.java#L394-L397
2. Deleting an 'inner'  object renders the outer objects invalid and they must be deleted. This is guaranteed by the foreign key constraints. Here is an example: https://github.com/DependencyTrack/hyades-apiserver/blob/279450cf3af0bec1ff23e712a2168b5d3508250b/src/main/resources/migration/changelog-v5.6.0.xml#L269-L272